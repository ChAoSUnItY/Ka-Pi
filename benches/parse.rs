use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ka_pi::parse::to_class;

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("parse main", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Main.class"
    )))));

    c.bench_function("parse enum", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Enum.class"
    )))));

    c.bench_function("parse record", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/Record.class"
    )))));

    c.bench_function("parse visible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/VisibleAnnotation.class"
    )))));

    c.bench_function("parse invisible annotation", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/InvisibleAnnotation.class"
    )))));

    c.bench_function("parse annotation target", |b| b.iter(|| to_class(black_box(include_bytes!(
        "../compiled_source/out/production/compiled_source/AnnotationTarget.class"
    )))));
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
