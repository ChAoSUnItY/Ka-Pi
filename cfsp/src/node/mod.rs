//! `node` module contains all specification-described data structures from
//! [The JVM Specification - Java SE 20 Edition](https://docs.oracle.com/javase/specs/jvms/se20/jvms20.pdf).
//!
//! Some of the structures are made to be user-friendly, at the same time makes it much more
//! straightforward to use.

pub mod access_flag;
pub mod attribute;
pub mod class;
pub mod constant;
mod error;
pub mod field;
pub mod method;
pub mod opcode;
pub mod signature;
