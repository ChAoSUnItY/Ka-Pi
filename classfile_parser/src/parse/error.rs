use std::io;

use thiserror::Error;

use crate::node::opcode::Opcode;

pub type ParseResult<T> = Result<T, ParseError>;

#[derive(Debug, Error)]
pub enum ParseError {
    #[error(transparent)]
    Io(#[from] io::Error),
    #[error("expected [0xCA, 0xFE, 0xBA, 0xBE] at the header of class file but got {0:?}")]
    MismatchedMagicNumber([u8; 4]),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBoundUsize(&'static str, Vec<&'static str>, usize),
    #[error("attempt to match {0} in {1:?} but got '{2}'")]
    MatchOutOfBoundChar(&'static str, Vec<char>, char),
    #[error("attempt to match opcode but got opcode value {0}")]
    MatchOutOfBoundOpcode(u8),
    #[error("attempt to match sub opcode in wide opcode but got {0}, while {1:?} is allowed as sub opcode of wide opcode")]
    MatchOutOfBoundWideOpcode(u8, Vec<Opcode>),
    #[error("the parse for segment is finished but still remains {0} bytes")]
    Remains(usize),
    #[error("expected {1:?} but got '{0}'")]
    MismatchedCharacter(char, Vec<char>),
    #[error("attempt to parse {0} but is incomplete")]
    OutOfBound(&'static str),
}
