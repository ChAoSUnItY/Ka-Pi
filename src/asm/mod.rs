pub use handle::Handle;

mod attribute;
pub mod byte_vec;
pub mod class;
mod constants;
mod edge;
mod frame;
mod handle;
mod handler;
mod label;
/// The JVM opcodes, access flags and array type codes. This interface does not define all the JVM
/// opcodes because some opcodes are automatically handled. For example, the xLOAD and xSTORE opcodes
/// are automatically replaced by xLOAD_n and xSTORE_n opcodes when possible. The xLOAD_n and
/// xSTORE_n opcodes are therefore not defined in this interface. Likewise for LDC, automatically
/// replaced by LDC_W or LDC2_W when necessary, WIDE, GOTO_W and JSR_W.
///
/// See <a href="https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-6.html">JVMS 6</a><br/>
/// **Author** Eric Bruneton<br/>
/// **Author** Eugene Kuleshov<br/>
pub mod opcodes;
mod symbol;
pub mod types;
pub mod record;
pub mod field;
pub mod method;
pub mod annotation;
pub mod signature;
