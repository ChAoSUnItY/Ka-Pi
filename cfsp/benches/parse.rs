use cfsp::parse::{to_class, ParsingOption};
use std::fs;
use std::io::Cursor;

use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, Throughput};

fn parsing_classes(c: &mut Criterion) {
    let mut g = c.benchmark_group("parse class");

    let compiled_source_folder =
        fs::read_dir("../compiled_source/out/production/compiled_source").unwrap();

    for entry in compiled_source_folder {
        let entry = entry.as_ref().unwrap().path();
        let class_name = entry.file_name().unwrap().to_str().unwrap();
        let class_bytes = fs::read(&entry).unwrap();

        g.throughput(Throughput::Bytes(class_bytes.len() as u64));
        g.bench_with_input(
            BenchmarkId::from_parameter(class_name),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default().parse_attribute())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );

        g.bench_with_input(
            BenchmarkId::from_parameter(format!("{class_name} (No attribute parsing)")),
            &class_bytes,
            |b, bytes| {
                b.iter_batched(
                    || Cursor::new(bytes),
                    |mut cursor| {
                        to_class(&mut cursor, ParsingOption::default())
                            .expect("Parsing fails on benchmarking");
                    },
                    BatchSize::SmallInput,
                );
            },
        );
    }
}

criterion_group!(benches, parsing_classes);
criterion_main!(benches);
