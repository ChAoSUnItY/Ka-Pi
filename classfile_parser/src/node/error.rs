use crate::parse::ParseError;
use thiserror::Error;

pub type NodeResResult<T> = Result<T, NodeResError>;

#[derive(Debug, Error)]
pub enum NodeResError {
    #[error(transparent)]
    ParseFail(#[from] ParseError),
    #[error("failed to parse bytes to string in constant Utf8, given bytes: {0:?}")]
    StringParseFail(Box<[u8]>),
    #[error("unknown constant reference #{0}")]
    UnknownConstantReference(u16),
    #[error("attempt to match {0} in {1:?} but got {2}")]
    MatchOutOfBound(&'static str, Vec<&'static str>, usize),
    #[error("expected referenced constant {0} at #{1} but got {2}")]
    MismatchReferenceConstant(&'static str, u16, &'static str),
    #[error("character {0} cannot be converted into wildcard")]
    InvalidWildcard(char),
    #[error("character {0} cannot be converted into base type")]
    InvalidBaseType(char),
}
