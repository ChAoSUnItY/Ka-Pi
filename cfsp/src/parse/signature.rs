use std::iter::Peekable;
use std::str::Chars;

use crate::node::signature::{
    ArrayType, BaseType, ClassType, FormalTypeParameter, ReferenceType, Signature, SignatureType,
    ThrowsType, TypeArgument, TypeVariable, WildcardIndicator,
};
use crate::parse::error::{ParseError, ParseResult};

const EXCLUDED_IDENTIFIER_CHARACTERS: &'static str = ".;[/<>:";

pub fn class_signature(input: &str) -> ParseResult<Signature> {
    let mut input = input.chars().peekable();
    let formal_type_parameters = formal_types(&mut input)?;
    let super_class = class_type_signature(&mut input)?;
    let mut interfaces = Vec::new();

    while let Some(char) = input.peek() {
        if *char == 'L' {
            interfaces.push(class_type_signature(&mut input)?);
        } else {
            break;
        }
    }

    if input.peek().is_some() {
        let remain = input.collect::<Vec<_>>();

        Err(ParseError::Remains(remain.len()))
    } else {
        Ok(Signature::Class {
            formal_type_parameters,
            super_class,
            interfaces,
        })
    }
}

pub fn field_signature(input: &str) -> ParseResult<Signature> {
    let mut input = input.chars().peekable();
    let field_type = reference_type(&mut input)?;

    if input.peek().is_some() {
        let remain = input.collect::<Vec<_>>();

        Err(ParseError::Remains(remain.len()))
    } else {
        Ok(Signature::Field { field_type })
    }
}

pub fn method_signature(input: &str) -> ParseResult<Signature> {
    let mut input = input.chars().peekable();
    let formal_type_parameters = formal_types(&mut input)?;

    assert_char(input.next(), '(')?;

    let mut parameter_types = Vec::new();

    while let Some(char) = input.peek() {
        if *char != ')' {
            parameter_types.push(java_type_signature(&mut input)?);
        } else {
            break;
        }
    }

    assert_char(input.next(), ')')?;

    let return_type = result(&mut input)?;
    let mut exception_types = Vec::new();

    while let Some(char) = input.peek() {
        if *char == '^' {
            exception_types.push(throws_signature(&mut input)?);
        } else {
            break;
        }
    }

    if input.peek().is_some() {
        let remain = input.collect::<Vec<_>>();

        Err(ParseError::Remains(remain.len()))
    } else {
        Ok(Signature::Method {
            formal_type_parameters,
            parameter_types,
            return_type,
            exception_types,
        })
    }
}

fn formal_types(input: &mut Peekable<Chars>) -> ParseResult<Vec<FormalTypeParameter>> {
    assert_char(input.next(), '<')?;
    let mut formal_types = Vec::new();

    while let Some(char) = input.peek() {
        if *char != '>' {
            formal_types.push(formal_type(input)?);
        } else {
            break;
        }
    }

    assert_char(input.next(), '>')?;

    Ok(formal_types)
}

fn formal_type(input: &mut Peekable<Chars>) -> ParseResult<FormalTypeParameter> {
    let parameter_name = identifier(input)?;
    let class_bound = class_bound(input)?;
    let mut interface_bounds = Vec::new();

    while let Some(char) = input.peek() {
        if *char == ':' {
            interface_bounds.push(interface_bound(input)?);
        } else {
            break;
        }
    }

    Ok(FormalTypeParameter {
        parameter_name,
        class_bound,
        interface_bounds,
    })
}

fn result(input: &mut Peekable<Chars>) -> ParseResult<SignatureType> {
    if let Ok(java_type_signature) = java_type_signature(input) {
        Ok(java_type_signature)
    } else if let Ok(void_descriptor) = void_descriptor(input) {
        Ok(SignatureType::BaseType(void_descriptor))
    } else {
        Err(ParseError::OutOfBound("return type"))
    }
}

fn throws_signature(input: &mut Peekable<Chars>) -> ParseResult<ThrowsType> {
    assert_char(input.next(), '^')?;

    if let Ok(class_type_signature) = class_type_signature(input) {
        Ok(ThrowsType::Class(class_type_signature))
    } else if let Ok(type_variable) = type_variable(input) {
        Ok(ThrowsType::TypeVariable(type_variable))
    } else {
        Err(ParseError::OutOfBound("throws signature"))
    }
}

fn class_bound(input: &mut Peekable<Chars>) -> ParseResult<Option<ClassType>> {
    assert_char(input.next(), ':')?;

    match input.peek() {
        Some(char) if *char == 'L' => Ok(Some(class_type_signature(input)?)),
        _ => Ok(None),
    }
}

fn interface_bound(input: &mut Peekable<Chars>) -> ParseResult<ClassType> {
    assert_char(input.next(), ':')?;

    class_type_signature(input)
}

fn java_type_signature(input: &mut Peekable<Chars>) -> ParseResult<SignatureType> {
    if let Ok(base_type) = base_type(input) {
        Ok(SignatureType::BaseType(base_type))
    } else if let Ok(reference_type) = reference_type(input) {
        Ok(SignatureType::ReferenceType(reference_type))
    } else {
        Err(ParseError::OutOfBound("java type signature"))
    }
}

fn base_type(input: &mut Peekable<Chars>) -> ParseResult<BaseType> {
    let Some(char) = input.peek()
    else {
        return Err(ParseError::OutOfBound("base type"));
    };

    for base_type_char in "ZBSIJFD".chars() {
        if *char == base_type_char {
            let char = input.next().unwrap();

            return BaseType::try_from(char).map_err(|_| {
                ParseError::MismatchedCharacter(char, vec!['Z', 'B', 'S', 'I', 'J', 'F', 'D'])
            });
        }
    }

    Err(ParseError::MismatchedCharacter(
        *char,
        vec!['Z', 'B', 'S', 'I', 'J', 'F', 'D'],
    ))
}

fn void_descriptor(input: &mut Peekable<Chars>) -> ParseResult<BaseType> {
    assert_char(input.next(), 'V')?;

    Ok(BaseType::Void)
}

fn reference_type(input: &mut Peekable<Chars>) -> ParseResult<ReferenceType> {
    match input.peek() {
        Some(char) if *char == 'L' => Ok(ReferenceType::Class(class_type_signature(input)?)),
        Some(char) if *char == 'T' => Ok(ReferenceType::TypeVariable(type_variable(input)?)),
        Some(char) if *char == '[' => Ok(ReferenceType::Array(array_type(input)?)),
        Some(char) => Err(ParseError::MismatchedCharacter(*char, vec!['L', 'T', '['])),
        None => Err(ParseError::OutOfBound("reference type")),
    }
}

fn class_type_signature(input: &mut Peekable<Chars>) -> ParseResult<ClassType> {
    assert_char(input.next(), 'L')?;

    let (package_path, class_name) = package_specifier_and_class_type(input)?;
    let type_arguments = type_arguments(input)?;
    let mut inner_classes = Vec::new();

    while let Some(char) = input.peek() {
        if *char != ';' {
            inner_classes.push(class_type_signature_suffix(input)?);
        } else {
            break;
        }
    }

    assert_char(input.next(), ';')?;

    Ok(ClassType {
        package_path,
        class_name,
        type_arguments,
        inner_classes,
    })
}

fn package_specifier_and_class_type(input: &mut Peekable<Chars>) -> ParseResult<(String, String)> {
    let mut class_path_builder = Vec::with_capacity(3);

    loop {
        class_path_builder.push(identifier(input)?);

        match input.peek() {
            Some(builder) if *builder == '/' => {
                class_path_builder.push(input.next().unwrap().to_string());
                continue;
            }
            _ => break,
        }
    }

    let (class_type, package_specifier) = if class_path_builder.len() == 1 {
        (String::new(), class_path_builder[0].clone())
    } else if class_path_builder.len() == 0 {
        return Err(ParseError::OutOfBound("package specifier and class type"));
    } else {
        class_path_builder
            .split_last()
            .map(|(class_type, package_specifier)| {
                (class_type.to_string(), package_specifier.join(""))
            })
            .unwrap()
    };

    Ok((package_specifier, class_type))
}

fn class_type_signature_suffix(
    input: &mut Peekable<Chars>,
) -> ParseResult<(String, Vec<TypeArgument>)> {
    assert_char(input.next(), '.')?;

    let class_type = identifier(input)?;
    let type_arguments = type_arguments(input)?;

    Ok((class_type, type_arguments))
}

fn type_arguments(input: &mut Peekable<Chars>) -> ParseResult<Vec<TypeArgument>> {
    match input.peek() {
        Some(char) if *char == '<' => {
            assert_char(input.next(), '<')?;

            let mut type_arguments = Vec::new();

            while let Some(char) = input.peek() {
                if *char != '>' {
                    type_arguments.push(type_argument(input)?);
                } else {
                    break;
                }
            }

            assert_char(input.next(), '>')?;

            Ok(type_arguments)
        }
        _ => Ok(Vec::new()),
    }
}

fn type_argument(input: &mut Peekable<Chars>) -> ParseResult<TypeArgument> {
    match input.peek() {
        Some(char) if *char == '+' || *char == '-' => {
            let wildcard_indicator = WildcardIndicator::try_from(char).unwrap();
            let bounded_type = reference_type(input)?;

            Ok(TypeArgument::Bounded {
                wildcard_indicator,
                bounded_type,
            })
        }
        Some(char) if *char == '*' => Ok(TypeArgument::Wildcard),
        Some(char) => Err(ParseError::MismatchedCharacter(*char, vec!['+', '-', '*'])),
        None => Err(ParseError::OutOfBound("type_argument")),
    }
}

fn type_variable(input: &mut Peekable<Chars>) -> ParseResult<TypeVariable> {
    assert_char(input.next(), 'T')?;

    let identifier = identifier(input)?;

    assert_char(input.next(), ';')?;

    Ok(TypeVariable(identifier))
}

fn array_type(input: &mut Peekable<Chars>) -> ParseResult<ArrayType> {
    assert_char(input.next(), '[')?;

    let inner_type = java_type_signature(input)?;

    Ok(ArrayType(Box::new(inner_type)))
}

fn identifier(input: &mut Peekable<Chars>) -> ParseResult<String> {
    let mut identifier_builder = String::with_capacity(8);

    while let Some(char) = input.peek() {
        if !EXCLUDED_IDENTIFIER_CHARACTERS.contains(*char) {
            identifier_builder.push(input.next().unwrap());
        } else {
            break;
        }
    }

    Ok(identifier_builder)
}

#[inline]
fn assert_char(char: Option<char>, expected: char) -> ParseResult<char> {
    match char {
        Some(char) => {
            if char == expected {
                Ok(char)
            } else {
                Err(ParseError::MismatchedCharacter(char, vec![expected]))
            }
        }
        None => Err(ParseError::OutOfBound("signature")),
    }
}
