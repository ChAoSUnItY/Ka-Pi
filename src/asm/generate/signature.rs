use std::collections::VecDeque;

use crate::asm::node::signature::{BaseType, Wildcard};
use crate::asm::visitor::signature::{
    ClassSignatureVisitor, FieldSignatureVisitor, FormalTypeParameterVisitable,
    FormalTypeParameterVisitor, MethodSignatureVisitor, TypeVisitor,
};

/// A default implementation of class signature writer.
///
/// This is commonly used in class file generation.
#[derive(Debug, Default)]
pub struct ClassSignatureWriter {
    signature_builder: String,
    has_formal: bool,
}

impl ClassSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
            has_formal: false,
        }
    }
}

impl ToString for ClassSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

impl ClassSignatureVisitor for ClassSignatureWriter {
    fn visit_super_class(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }

    fn visit_interface(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }
}

impl FormalTypeParameterVisitable for ClassSignatureWriter {
    //noinspection DuplicatedCode
    fn visit_formal_type_parameter(
        &mut self,
        name: &str,
    ) -> Box<dyn FormalTypeParameterVisitor + '_> {
        if !self.has_formal {
            self.has_formal = true;
            self.signature_builder.push('<');
        }

        self.signature_builder.push_str(name);

        Box::new(FormalTypeParameterWriter::new(&mut self.signature_builder))
    }
}

/// A default implementation of field signature writer.
///
/// This is commonly used in class file generation.
#[derive(Debug, Default)]
pub struct FieldSignatureWriter {
    signature_builder: String,
}

impl FieldSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
        }
    }
}

impl ToString for FieldSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

impl FieldSignatureVisitor for FieldSignatureWriter {
    fn visit_field_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeWriter::new(&mut self.signature_builder))
    }
}

#[derive(Debug, Default)]
pub struct MethodSignatureWriter {
    signature_builder: String,
    has_formal: bool,
    has_parameters: bool,
}

/// A default implementation of method signature writer.
///
/// This is commonly used in class file generation.
impl MethodSignatureWriter {
    /// Reserve capacity for builder to build.
    ///
    /// This is useful when mass appending is required.
    pub fn with_capacity(size: usize) -> Self {
        Self {
            signature_builder: String::with_capacity(size),
            has_formal: false,
            has_parameters: false,
        }
    }
}

impl ToString for MethodSignatureWriter {
    fn to_string(&self) -> String {
        self.signature_builder.clone()
    }
}

impl MethodSignatureVisitor for MethodSignatureWriter {
    fn visit_parameter_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        if !self.has_parameters {
            self.has_parameters = true;
            self.signature_builder.push('(');
        }

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }

    fn visit_return_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        if self.has_formal {
            self.has_formal = false;
            self.signature_builder.push('>');
        }

        if self.has_parameters {
            self.has_parameters = false;
        } else {
            self.signature_builder.push('(');
        }

        self.signature_builder.push(')');

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }

    fn visit_exception_type(&mut self) -> Box<dyn TypeVisitor + '_> {
        self.signature_builder.push('^');

        Box::new(TypeWriter::new(&mut self.signature_builder))
    }
}

impl FormalTypeParameterVisitable for MethodSignatureWriter {
    //noinspection DuplicatedCode
    fn visit_formal_type_parameter(
        &mut self,
        name: &str,
    ) -> Box<dyn FormalTypeParameterVisitor + '_> {
        if !self.has_formal {
            self.has_formal = true;
            self.signature_builder.push('<');
        }

        self.signature_builder.push_str(name);

        Box::new(FormalTypeParameterWriter::new(&mut self.signature_builder))
    }
}

struct FormalTypeParameterWriter<'parent> {
    parent_builder: &'parent mut String,
}

impl<'parent> FormalTypeParameterWriter<'parent> {
    fn new(parent_builder: &'parent mut String) -> Self {
        Self { parent_builder }
    }
}

impl<'parent> FormalTypeParameterVisitor for FormalTypeParameterWriter<'parent> {
    fn visit_class_bound(&mut self) -> Box<dyn TypeVisitor + '_> {
        Box::new(TypeWriter::new(&mut self.parent_builder))
    }

    fn visit_interface_bound(&mut self) -> Box<dyn TypeVisitor + '_> {
        self.parent_builder.push(':');
        Box::new(TypeWriter::new(&mut self.parent_builder))
    }
}

struct TypeWriter<'parent> {
    parent_builder: &'parent mut String,
    type_arg_stack: VecDeque<bool>,
}

impl<'parent> TypeWriter<'parent> {
    fn new(parent_builder: &'parent mut String) -> Self {
        let mut type_arg_stack = VecDeque::with_capacity(64);

        type_arg_stack.push_back(true);

        Self {
            parent_builder,
            type_arg_stack,
        }
    }

    fn end_args(&mut self) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.parent_builder.push('>');
        }

        self.type_arg_stack.pop_front();
    }
}

impl<'parent> TypeVisitor for TypeWriter<'parent> {
    fn visit_base_type(&mut self, base_type: BaseType) {
        self.parent_builder.push(base_type.into());
    }

    fn visit_array_type(&mut self) {
        self.parent_builder.push('[');
    }

    fn visit_class_type(&mut self, name: &str) {
        self.parent_builder.push('L');
        self.parent_builder.push_str(name);
        self.parent_builder.push(';');
        self.type_arg_stack.push_front(false);
    }

    fn visit_inner_class_type(&mut self, name: &str) {
        self.end_args();
        self.parent_builder.push('.');
        self.parent_builder.push_str(name);
        self.type_arg_stack.push_front(false);
    }

    fn visit_type_variable(&mut self, name: &str) {
        self.parent_builder.push('T');
        self.parent_builder.push_str(name);
        self.parent_builder.push(';');
    }

    fn visit_type_argument(&mut self) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.type_arg_stack[0] = true;
            self.parent_builder.push('<');
        }

        self.parent_builder.push('*');
    }

    fn visit_type_argument_wildcard(&mut self, wildcard: Wildcard) {
        if self.type_arg_stack.front().map_or(false, |b| *b) {
            self.type_arg_stack[0] = true;
            self.parent_builder.push('<');
        }

        if wildcard != Wildcard::INSTANCEOF {
            self.parent_builder.push(wildcard.into());
        }
    }

    fn visit_end(&mut self) {
        self.end_args();
        self.parent_builder.push(';');
    }
}

#[cfg(test)]
mod test {
    use insta::assert_yaml_snapshot;

    use crate::asm::generate::signature::{
        ClassSignatureWriter, FieldSignatureWriter, MethodSignatureWriter,
    };
    use crate::asm::visitor::signature::{
        ClassSignatureVisitor, FieldSignatureVisitor, FormalTypeParameterVisitable,
        MethodSignatureVisitor,
    };

    #[test]
    fn test_class_signature_writer() {
        let mut writer = ClassSignatureWriter::default();

        writer
            .visit_formal_type_parameter(&"T".to_string())
            .visit_class_bound()
            .visit_class_type("java/lang/Object");

        writer
            .visit_super_class()
            .visit_class_type("java/lang/Object");
        writer
            .visit_interface()
            .visit_class_type("java/lang/Comparable");

        assert_yaml_snapshot!(writer.to_string());
    }

    #[test]
    fn test_field_signature_writer() {
        let mut writer = FieldSignatureWriter::default();

        writer
            .visit_field_type()
            .visit_class_type("java/lang/String");

        assert_yaml_snapshot!(writer.to_string());
    }

    #[test]
    fn test_method_signature_writer() {
        let mut writer = MethodSignatureWriter::default();

        writer
            .visit_formal_type_parameter(&"T".to_string())
            .visit_class_bound()
            .visit_class_type("java/lang/Object");

        writer
            .visit_parameter_type()
            .visit_class_type("java/lang/Object");
        writer
            .visit_return_type()
            .visit_class_type("java/lang/String");

        assert_yaml_snapshot!(writer.to_string());
    }
}
