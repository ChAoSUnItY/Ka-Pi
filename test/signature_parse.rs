use insta::assert_yaml_snapshot;

use ka_pi::asm::node::signature::{
    ClassType, FormalTypeParameter, ReferenceType, SignatureType, ThrowsType,
};
use ka_pi::asm::parse::{parse_class_signature, parse_field_signature, parse_method_signature};
use ka_pi::asm::visitor::signature::{
    ClassSignatureVisitor, FieldSignatureVisitor, FormalTypeParameterVisitable,
    FormalTypeParameterVisitor, MethodSignatureVisitor, SignatureVisitor, TypeVisitor,
};
use ka_pi::asm::visitor::Visitable;
use ka_pi::error::KapiResult;

#[test]
fn test_class_signature_with_generic() -> KapiResult<()> {
    let class_signature =
        parse_class_signature("<T:Ljava/lang/Object;>Ljava/lang/Object;Ljava/lang/Runnable;")?;

    assert_yaml_snapshot!(class_signature);

    Ok(())
}

#[test]
fn test_field_signature_object() -> KapiResult<()> {
    let field_signature = parse_field_signature("Ljava/lang/Object;")?;

    assert_yaml_snapshot!(field_signature);

    Ok(())
}

#[test]
fn test_field_signature_type_variable() -> KapiResult<()> {
    let field_signature = parse_field_signature("TT;")?;

    assert_yaml_snapshot!(field_signature);

    Ok(())
}

#[test]
fn test_method_signature_with_generic() -> KapiResult<()> {
    let method_signature = parse_method_signature(
        "<T:Ljava/lang/Object;>(Z[[ZTT;)Ljava/lang/Object;^Ljava/lang/Exception;",
    )?;

    assert_yaml_snapshot!(method_signature);

    Ok(())
}

// Visitor test
// Example usage of visitor: Remap specific type argument name.

#[derive(Default)]
struct Visitor {}

impl SignatureVisitor for Visitor {}

impl ClassSignatureVisitor for Visitor {
    type SCTV = Self;
    type ITV = Self;

    fn visit_super_class(&mut self, _: &mut ClassType) -> Self::SCTV {
        Self::default()
    }

    fn visit_interface(&mut self, _: &mut ClassType) -> Self::ITV {
        Self::default()
    }
}

impl FormalTypeParameterVisitable for Visitor {
    type FTPV = Self;

    fn visit_formal_type_parameter(
        &mut self,
        formal_type_parameter: &mut FormalTypeParameter,
    ) -> Self::FTPV {
        let FormalTypeParameter { parameter_name, .. } = formal_type_parameter;

        if parameter_name == "T" {
            *parameter_name = "V".to_string();
        } else if parameter_name == "R" {
            *parameter_name = "K".to_string();
        }

        Self::default()
    }
}

impl FieldSignatureVisitor for Visitor {
    type FTV = Self;

    fn visit_field_type(&mut self, _: &mut ReferenceType) -> Self::FTV {
        Self::default()
    }
}

impl MethodSignatureVisitor for Visitor {
    type PTV = Self;
    type RTV = Self;
    type ETV = Self;

    fn visit_parameter_type(&mut self, _: &mut SignatureType) -> Self::PTV {
        Self::default()
    }

    fn visit_return_type(&mut self, _: &mut SignatureType) -> Self::RTV {
        Self::default()
    }

    fn visit_exception_type(&mut self, _: &mut ThrowsType) -> Self::ETV {
        Self::default()
    }
}

impl FormalTypeParameterVisitor for Visitor {
    type CBTV = Self;
    type IBTV = Self;

    fn visit_class_bound(&mut self, _: &mut ClassType) -> Self::CBTV {
        Self::default()
    }

    fn visit_interface_bound(&mut self, _: &mut ClassType) -> Self::IBTV {
        Self::default()
    }
}

impl TypeVisitor for Visitor {}

#[test]
fn test_signature_visitor() -> KapiResult<()> {
    let mut class_signature =
        parse_class_signature("<T:Ljava/lang/Object;R:>Ljava/lang/Object;Ljava/lang/Runnable;")?;

    class_signature.visit(&mut Visitor::default());

    assert_yaml_snapshot!(class_signature);

    Ok(())
}
