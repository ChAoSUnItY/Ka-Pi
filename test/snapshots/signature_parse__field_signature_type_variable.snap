---
source: test/signature_parse.rs
expression: field_signature
---
Field:
  field_type:
    TypeVariable: T

